File: WoofIOS/SwiftPackages/Presentation/Favorites/Sources/PresentationFavorites/ViewModel/FavoritesViewModel.swift
import Combine
import DI
import Models
import PresentationCommon
import UseCase

@Factory
public final class FavoritesViewModel: StateDerivingViewModel<FavoritesViewModel.State, Never> {
    private let cellMapper: DogDomainToFavoriteCellUIModelMapper

    init(
        cellMapper: DogDomainToFavoriteCellUIModelMapper,
        getFavoriteDogsUseCase: GetFavoriteDogsUseCase
    ) {
        self.cellMapper = cellMapper
        super.init(
            initialValue: .empty,
            source: getFavoriteDogsUseCase()
                .map { dogs in
                    switch dogs.isEmpty {
                    case true: State.empty
                    case false: State.loaded(cells: dogs.map { cellMapper(dog: $0) })
                    }
                }
                .eraseToAnyPublisher()
        )
    }
}

extension FavoritesViewModel {
    public enum State: Equatable {
        case empty
        case loaded(cells: [FavoriteCellUIModel])
    }
}

File: WoofAndroid/presentation/favorites/src/main/kotlin/com/perrystreet/woof/presentation/favorites/viewmodel/FavoritesViewModel.kt
package com.perrystreet.woof.presentation.favorites.viewmodel

import com.perrystreet.woof.presentation.common.rx.UiObservable
import com.perrystreet.woof.presentation.common.rx.UiObservable.Companion.asUiObservable
import com.perrystreet.woof.presentation.common.viewmodel.StateDerivingViewModel
import com.perrystreet.woof.presentation.favorites.mapper.DogDomainToFavoriteCellUIModelMapper
import com.perrystreet.woof.presentation.favorites.uimodel.FavoriteCellUIModel
import com.perrystreet.woof.usecase.favorites.GetFavoriteDogsUseCase
import org.koin.android.annotation.KoinViewModel

@KoinViewModel
class FavoritesViewModel(
    private val cellMapper: DogDomainToFavoriteCellUIModelMapper,
    getFavoriteDogsUseCase: GetFavoriteDogsUseCase,
) : StateDerivingViewModel<FavoritesViewModel.State>() {

    sealed class State {
        data object Empty : State()

        data class Loaded(val cells: List<FavoriteCellUIModel>) : State()
    }

    override val state: UiObservable<State> = getFavoriteDogsUseCase()
        .map { dogs ->
            when (dogs.isEmpty()) {
                true -> State.Empty
                false -> State.Loaded(cells = dogs.map { cellMapper(it) })
            }
        }
        .asUiObservable(State.Empty)
}
